pta: add Pseudo TA for auxiliary crypto services

igoropaniuk opened the pull request:
AOSP Keymaster implementations need to securely mix the provided entropy into
their pool, which also must contain internally-generated entropy from a
hardware random number generator.

Add PTA, which can add caller-provided entropy to the default RNG pool.

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented:
Even though pseudo TAs are quite cheap to add, it's not for free. We should consider that at some point we'll need to consolidate some PTAs. It doesn't have to start here, but the callers of this TA should be aware that this interface can change.

Last time we talked about this, a system PTA was suggested instead.

igoropaniuk commented:
@jenswi-linaro 

> Last time we talked about this, a system PTA was suggested instead.
Ok, will fix now

and btw, should this particular `addEntropyCall` call be covered in optee_test (I'm 100 sure that it should, but it's better to ask :) ).

jenswi-linaro commented:
> and btw, should this particular addEntropyCall call be covered in optee_test (I'm 100 sure that it should, but it's better to ask :) ).

We can add something when the keystorage TA starts to use that call.

etienne-lms commented on core/arch/arm/pta/crypto_aux.c:
```diff
@@ -0,0 +1,75 @@
+/* SPDX-License-Identifier: BSD-2-Clause
```
minor: check expected format

etienne-lms replied:
yes, I though the C source way was `// SPDX-License-Identifier: xxx`

igoropaniuk replied:
@etienne-lms there are a bunch of files with C-style comments, for example. So which one to choose? :)
https://github.com/OP-TEE/optee_os/blob/master/core/include/io.h

jforissier replied:
@igoropaniuk see rules 1.2.1 and 1.2.2 in https://github.com/OP-TEE/optee_os/blob/master/documentation/copyright_and_license_headers.rst.

etienne-lms replied:
check these guidelines: https://github.com/OP-TEE/optee_os/blob/master/documentation/copyright_and_license_headers.rst#new-source-files.

etienne-lms commented on core/arch/arm/pta/crypto_aux.c:
```diff
@@ -0,0 +1,75 @@
+		    * depends on those two flags we add them explicitly
+		    * too.
+		    */
+		   .flags = PTA_DEFAULT_FLAGS | TA_FLAG_SINGLE_INSTANCE |
```
single instance flag is meaningless for pTAs: there are all single instance.

igoropaniuk replied:
@etienne-lms fixed

etienne-lms commented on core/arch/arm/pta/crypto_aux.c:
```diff
@@ -0,0 +1,75 @@
+	struct tee_ta_session *s;
+
+	/* Check that we're called from a user TA */
+	s = tee_ta_get_calling_session();
```
why forbidding non-secure to feed entropy?

igoropaniuk replied:
The purpose of this pTA is to extend GP Core API, which doesn't define any call for re-seeding entropy pool.

@jenswi-linaro could you please share your thoughts regarding non-secure calls to this pTA?

jenswi-linaro replied:
This PTA is only supposed to be called from a secure TA. Actually I wouldn't mind if there was an extra parameter to the call indicating if the origin of the entropy is secure or non-secure even if we will ignore that for now.

igoropaniuk commented:
All comments address,
also added additional check for a seed size (Fortuna PRNG requires seed to be <= 32 bytes)

Added xtest test case https://github.com/OP-TEE/optee_test/pull/283

jenswi-linaro commented on core/arch/arm/pta/system.c:
```diff
@@ -0,0 +1,69 @@
+	entropy_sz = params[0].memref.size;
+
+	/* Fortuna PRNG requires seed <= 32 bytes */
+	if (!entropy_sz && entropy_sz > 32)
```
I think it would be more friendly to just ignore the extra bytes.

jenswi-linaro commented on ta/ta.mk:
```diff
@@ -16,6 +16,10 @@ ifeq ($(CFG_TA_MBEDTLS_SELF_TEST),y)
 $(sm)-platform-cppflags += -DMBEDTLS_SELF_TEST
 endif
 
+ifeq ($(CFG_SYSTEM_PTA),y)
```
These lines shouldn't be needed, CFG-variables should propagate automatically via `ta-mk-file-export-vars-$(sm)`

jenswi-linaro commented on ta/ta.mk:
```diff
@@ -29,6 +33,10 @@ ta-mk-file-export-vars-$(sm) += CFG_SECURE_DATA_PATH
 ta-mk-file-export-vars-$(sm) += CFG_TA_MBEDTLS_SELF_TEST
 ta-mk-file-export-vars-$(sm) += CFG_TA_MBEDTLS
 
+ifeq ($(CFG_SYSTEM_PTA),y)
```
No need to test for `y`, the actual value is propagated

igoropaniuk commented on core/arch/arm/pta/system.c:
```diff
@@ -25,9 +25,12 @@ static TEE_Result system_rng_reseed(struct tee_ta_session *s __unused,
+	if (!entropy_sz)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	entropy_sz =
```
@jenswi-linaro,

sorry, thought that `optee_os/lib/libutils/ext/include/util.h` is supposed to be included only within TAs (+libutee etc.), but not within core.

Will fix now

igoropaniuk commented on lib/libutee/include/pta_system.h:
```diff
@@ -0,0 +1,28 @@
+ */
+#define PTA_SYSTEM_ADD_RNG_ENTROPY	0
+
+#define MAX_ENTROPY_IN			32u
```
fixed

igoropaniuk commented:
Tag applied, rebased and ready for merge. Thanks!
